### ISSUE TITLE
Add CollectionsMarshal.AsSpan support for Queue<T>

### ISSUE THREAD

**Issue by Symbai** [dotnet/runtime#1679]
CollectionsMarshal only supports `List<T>` but not `Queue<T>`. Queue also has an array property, adding support for it should is trivial and the benefits / purposes are the same than for List. 

https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.cs#L202-L205

https://github.com/dotnet/runtime/blob/adc91a7f52ccdb3156f244fffda89be26282669c/src/libraries/System.Collections/src/System/Collections/Generic/Queue.cs#L29

**Comment by stephentoub** [comment]
I don't think the benefits are the same, especially since you need additional information for the queue to be useful in many situations, e.g. the index of the first element in the queue. Lists are also much more prevalent than queue.

Can you share where, why, and how you'd use this?

**Comment by Symbai** [comment]
Queue is often used in situations where you add many items and then remove many items in the same order and order is no supposed to change (or order doesn't matter at all). It clearly outperforms a List in these cases performance wise.

I have an operation where I'm adding an unknown amount of items to a collection (thus I cannot use an array). In some cases it can be more than 15 million. Using `Queue<T>` performs MUCH faster than using `List<T>` and a bit faster than using `Stack<T>`. However there are certain APIs that only accept arrays, especially Span related APIs. 

Here is one of my code where I'm writing this large collection of structs to disk as fast as possible. Instead of writing struct by struct which can take a very long time I am simply writing the whole collection to disk. This is super fast and enough for my purpose. Sadly it creates a full copy of this large collection in memory.
```
//elements in queue >= 15,000,000
using var fs = new FileStream(filePath, FileMode.OpenOrCreate, FileAccess.ReadWrite);
Span<MyStruct> span = queue.ToArray();
fixed (void* ptr = &Unsafe.As<MyStruct, byte>(ref span[0]))
{
    var bytes = new Span<byte>(ptr, span.Length * Unsafe.SizeOf<MyStruct>());
    fs.Write(bytes);
}
```
For this code to work with a Queue I have to call ToArray first which allocations a copy in memory only to write it to the disk. I don't even need this copy at all but it is unavoidable currently. If I could use CollectionsMarshal like this:

```
//elements in queue >= 5,000,000
using var fs = new FileStream(filePath, FileMode.OpenOrCreate, FileAccess.ReadWrite);
Span<MyStruct> span = CollectionsMarshal.AsSpan(queue);
fixed (void* ptr = &Unsafe.As<MyStruct, byte>(ref span[0]))
{
    var bytes = new Span<byte>(ptr, span.Length * Unsafe.SizeOf<MyStruct>());
    fs.Write(bytes);
}
```
I would save the additional allocation. But sadly it only accepts `List<T>` so I would have to use a List instead of Queue, which eliminates the benefit I had when using Queue. I would no longer create a copy of such a large collection, which is great but the performance when adding these elements is dropped by half, which is not so great.

And sadly there is no easy way to write a similar API myself as far as I know. I would have to create my on Queue class which exposes an array while adding it on your end is rather trivial. 

So to sum it up:

Where: In all situations where the developer has strong reasons to not using List but Queue (or another collection)
Why: There are certain APIs (especially Span related) that only accepts arrays and Queue has this array but it's not exposed.
How: Depends on the situation but I exposed mine above.

**Comment by stephentoub** [comment]
> If I could use CollectionsMarshal like this:

That would not work correctly.  There is no guarantee, and in fact it's unlikely, that all of the data in the queue will be contiguous.  Consider this:
```C#
var q = new Queue<int>();
q.Enqueue(1);
q.Dequeue();
q.Enqueue(2);
q.Dequeue();
q.Enqueue(3);
q.Dequeue();
q.Enqueue(4);
q.Enqueue(5);
Debugger.Break();
```
If AsSpan were to return you a span for the array inside of `Queue<T>`, you'd have span[0] == 5, span[3] == 4, and span[1] and span[2] would be garbage:
![image](https://user-images.githubusercontent.com/2642209/72281303-3f6d7e00-3608-11ea-8c01-7f1cf9225146.png)

**Comment by Symbai** [comment]
Oh, that's interesting. But in my example I don't dequeue. I only enqueue then write them to disk so it would be contiguous? Consider the API is "not so safe to use" anyway because it says i shouldn't be used while adding / removing to List, it might be "okay" to add a similar note for queues "Should not be used when items have been dequeued before"?

**Comment by stephentoub** [comment]
If you never dequeue, why are you using a `Queue<T>`?  I see you wrote `Using Queue<T> performs MUCH faster than using List<T> and a bit faster than using Stack<T>`, but that makes very little sense to me.  Can you please share your benchmark?  I just tried, and get this:

```C#
[Benchmark]
public int AddList()
{
    var list = new List<int>();
    for (int i = 0; i < 1_000_000; i++)
    {
        list.Add(i);
    }
    return list.Count;
}

[Benchmark]
public int AddQueue()
{
    var queue = new Queue<int>();
    for (int i = 0; i < 1_000_000; i++)
    {
        queue.Enqueue(i);
    }
    return queue.Count;
}
```

|   Method |     Mean |     Error |    StdDev |     Gen 0 |     Gen 1 |     Gen 2 | Allocated |
|--------- |---------:|----------:|----------:|----------:|----------:|----------:|----------:|
|  AddList | 3.612 ms | 0.0499 ms | 0.0416 ms | 1996.0938 | 1996.0938 | 1996.0938 |      8 MB |
| AddQueue | 5.303 ms | 0.0813 ms | 0.0761 ms | 1992.1875 | 1992.1875 | 1992.1875 |      8 MB |

I don't think we should add such an AsSpan method for queue.  Saying it's only usable if dequeue has never been used makes it even more a pit of failure and severely limits its usefulness, which IMO is already low.

**Comment by Symbai** [comment]
In your benchmark it is about the same (funny though Queue is still 0.4ms faster on my end) but when I switch Queue to list in my method it is noticeably slower with an additional allocation of 0.04GB. I also don't dequeue when writing to disk, which I only showed you because you asked me for a situation where exactly  I am going to need this CollectionsMarshal for queue. The same method also runs without storing to disk and dequeue it in another method instead.

However I didn't know about using dequeue is not staying in sync with the array. That is a reasonably argument to not implement it. Guess I have no other way than copy the queue code and expose the array to get rid of this unnecessary allocation.
